pkg/archive/archive.go

```diff
@@ -148,12 +149,18 @@ func DecompressStream(archive io.Reader) (io.ReadCloser, error) {
 		readBufWrapper := p.NewReadCloserWrapper(buf, bz2Reader)
 		return readBufWrapper, nil
 	case Xz:
-		xzReader, err := xzDecompress(buf)
+		xzReader, chdone, err := xzDecompress(buf)
 		if err != nil {
 			return nil, err
 		}
 		readBufWrapper := p.NewReadCloserWrapper(buf, xzReader)
-		return readBufWrapper, nil
+		return ioutils.NewReadCloserWrapper(readBufWrapper, func() error {
+			go func() {
```

Comment:
you're right; the reason for deadlock is xz pending on writing the pipe; the separate goroutine waiting on channel to close didn't actually fix, it moved the logic to a goroutine to waiting on xz,

```
docker-daemon cat the-tarball.xz | xz -d -c -q | docker-untar /path/to/... (aufs )
````

while now I have another fix, the debug log shows it passed `readBufWrapper.Close()`, will cleanup and push the fix again

```
WARN[0017] chdone (chan struct {})(0xc8204cfb60) closed    <= after cmd.Wait and close the channel
DEBU[0017] Couldn't untar to /home/lib-docker-v1.9.0-tmp/0.0/aufs/mnt/4e2714008091e46c345efb67a1948b66b498cd8adea8734afd6dc331d3dfd44a: Untar re-exec error: exit status 1: output: operation not supported 
DEBU[0017] CopyFileWithTar(/home/lib-docker-v1.9.0-tmp/0.0/tmp/docker-builder357741497/stage3-x32-20151004.tar.xz, /home/lib-docker-v1.9.0-tmp/0.0/aufs/mnt/4e2714008091e46c345efb67a1948b66b498cd8adea8734afd6dc331d3dfd44a/stage3-x32-20151004.tar.xz) 
WARN[0017] waiting channel (<-chan struct {})(0xc8204cfb60) then to close &ioutils.readCloserWrapper{Reader:(*io.PipeReader)(0xc820032838), closer:(func() error)(0x976a30)} 
WARN[0017] readBufWrapper (<-chan struct {})(0xc8204cfb60) closed now.
```